### DISCUSSION TITLE
Test Project Support

### DISCUSSION THREAD

**Discussion by joshmouch** [microsoft/aspire#3125] [General]
I'm trying to find the best way to add a test project (or DLL or EXE) to be run from an Aspire AppHost.  When using Visual Studio and debugging with F5, Visual Studio should also attach to the tests (and test project).

If I were using xUnit v3 (which has an EXE entry point instead of dll), I think I could just use `.AddProject("ExampleXunit3TestProject.csproj")`.  As far as I know, any environment variables defined by `.WithReference()` would remain intact while the tests are running.  I don't know that for sure, though, since xUnit v3 generates a `Main()` that, in turn, creates an `XunitConsoleRunner`.  That sounds like a new process to me (and therefore not attached to debugger), but still investigating. The author stated that `VSTestRunner` isn't available yet, which is the one that I think would attach the debugger to the new processes.  @bradwilson

Using xUnit v2, you have to use a test runner since there's no EXE to call.  xUnit has their own test runners (e.g. `xunit.runner.console`), but it'd be preferrable to use a test framework agnostic test runner, which means `vstest.exe`, or `dotnet test` or similar.  In my initial testing, I lose my environment variables with `dotnet test`, which means the tests don't have access to the modifications to environment from `.WithReference()`.  In other words, using `.AddExecutable()` with `dotnet test` results in tests not having access to `.WithReference()`.

In both of the above, it's unclear to me if VisualStudio would attach to the processes being spawned.  I guess that's up to how `.AddExecutable()` is implemented.  If it doesn't today, then it'd be great to do so or else add an option to do so.

Ideally, it'd be great to add a new `DistributedApplicationBuilder` extension `.AddDotNetCliCommand()` or `.AddTestProject()` specifically to add test projects with the outcome of running them with `dotnet test`.

I think this also paves the way to being able to run tests from Visual Studio (or Resharper) UI into a docker container.  To my knowledge, this is either not possible today, or not documented.  This is assuming Aspire is going to allow F5 to put projects into a container instead of in process (through a launchProfile, perhaps?), and that it'll automatically attach to the process in the container.  That or else use the trick that VisualStudio does with docker files where it mounts the EXE as a volume and just attaches to it locally.

**Comment by joshmouch** [comment]
Another thought here is that it should be possible to get a list of the environment variables being created and sent to the `.AddExecutable()`.  With this, I could pass the environment variables into `dotnet test` myself using command line arguments.  That's not a great workflow and doesn't attach the debugger, but it's a quick fix.

**Comment by davidfowl** [comment]
I don’t understand the scenario here at all. Aspire doesn’t build containers or run your projects in containers. Running tests from the orchestrator isn’t a scenario that has come up. Testing orchestrated code is something that we’re going to provide but the end goal there is to be able to test your application and its dependencies

**Reply by joshmouch** [reply]
The scenario is to add a test project or binary to the orchestration to test against the application and dependencies within their currently deployed environment.

1. If that "deployed" environment is "Visual Studio + F5" then attaching a debugger is needed.  
2. That deployed environment could be Azure Container Apps as described here: https://learn.microsoft.com/en-us/dotnet/aspire/deployment/overview.
3. Or the deployed environment could be anywhere as long as there is tooling that supports the generated manifest.
4. (future) Maybe eventually the "deployed" environment is Visual Studio + Test Explorer, but that sounds like the `vstest.exe` team needs to get involved. I imagine JetBrains will add support sooner.

When I speak of containers, I mean either:
1. Aspire-manifest-aware tooling doing the pushing into containers (e.g. Azure Container Apps)
2. or else the already-built-in Visual Studio container tooling (either a dockerfile or .NET 7+ SDK MSBuild container support).  

If #2, it looks something like:
```
var apiService = builder.AddProject<Projects.AspireApp1_Web>("webfrontend")

    // If Visual Studio added a DockerFile when asked to add container support
    // This sets "Docker" as the launch profile commandName (a.k.a debug target to run)
    // The docker file uses the debugging tricks described here: https://learn.microsoft.com/en-us/visualstudio/containers/container-build?view=vs-2022#debugging
    .WithLaunchProfile("DockerProfile") 

    // If Visual Studio set the MSBuild variable `EnableSdkContainerDebugging=true` when asked to add container support
    // This sets "SdkContainer" as the launch profile commandName (a.k.a debug target to run)
    .WithLaunchProfile("SdkContainerDebuggingProfile") 

    .WithReference(cache)
    .WithReference(apiService)
```

So, that said, the problem I'm posing is: what comes after that last code snippet to add and debug a test project, regardless of the deployed environment.

**Reply by davidfowl** [reply]
Maybe you could make this work with the pieces we have today (with lots of additional work), but it's not something we've thought of and frankly, not something we'd prioritize in the short term. None of the things you mention are how any of the orchestration works today, so I'm not sure where that leaves you.

> If that "deployed" environment is "Visual Studio + F5" then attaching a debugger is needed. (Maybe eventually this is Visual Studio + Test Explorer)

You don't run your tests orchestrated as part of the apphost in this scenario. The tests run locally against the a locally running orchestrated application. e.g. https://github.com/dotnet/aspire/blob/49751a3b0e6cb0540ededbb5870ef31f101932b2/tests/Aspire.Hosting.Testing.Tests/TestingBuilderTests.cs#L48-L55

> That deployed environment could be Azure Container Apps as described here: https://learn.microsoft.com/en-us/dotnet/aspire/deployment/overview.

Not sure what this means. Does it mean you deployed your application and are running end to end tests hitting the urls locally?

> Or the deployed environment could be anywhere as long as there is tooling that supports the generated manifest.

Needs more detail. How does it all hang together? Where are the tests running? How does it get the relevant information in order to execute the test? What are these tests doing?

I would start with a single end to end that you want to accomplish and work backwards to figure out what features are required to accomplish it.

Love the exploration here though!

**Reply by joshmouch** [reply]
> it's not something we've thought of and frankly, not something we'd prioritize in the short term
> ...
> You don't run your tests orchestrated as part of the apphost in this scenario. The tests run locally against the a locally running orchestrated application.

Really, all that we're talking about is letting an existing test project in on some of this sweet, sweet, sweet **Service Discovery action**.  Which is almost already possible with what's here... just have to figure out how.  The blocker is getting those generated environment variables down to the tests.  ...and then also not breaking the ability to debug tests in doing so.
And then the second thing is supporting the "SdkContainer" and "Docker" launch profiles, where Visual Studio is already doing most of the work.

---

In your example test code, above, that test is only useable up until that builder becomes a manifest and hits tooling.  It won't be useable targeted at a deployed environment.  

So then, you say: to reuse those tests, you configure the endpoints manually versus using Service Discovery and then abstract the HttpClient factory.  When running local, you can use the logical URL and the ServiceDiscovery `HttpHandler`, but when running remote, you use a physical URL and a normal `HttpHandler` stack.

Then the developer says: Fine, that was some extra work to find and configure those services without the service discovery, and I had to modify my test project to have a dependency on the orchestration project, but I did it.

But then he adds: Since I can only use the tests that target public endpoints unless they're deployed within the same environment, I have to create a second deployment to get the test binaries inside there.  And the deployment is completely different, since one uses Aspire tooling and one doesn't.  Now I'm writing and maintain two styles of doing it.  I was hoping it could all be one deployment.

**Reply by davidfowl** [reply]
> Really, all that we're talking about is letting an existing test project in on some of this sweet, sweet, sweet Service Discovery action

if that’s all you need then skip the other stuff. You don’t need to do anything other than get the urls from the deployment in order to hit your services. Of course this work just work if those urls are exposed to the public internet.

I’m not sure why you need orchestration or anything else really.

**Reply by bradwilson** [reply]
As mentioned [here](https://github.com/xunit/xunit/discussions/2902#discussioncomment-8894872), you can control the entry point in xUnit.net v3. The envisioned scenario was actually exactly like this: you want to set up an environment ahead of time before running tests. The fact that we push to a stand-alone executable by default is not strictly speaking necessary; our primary purpose here with saying tests are "stand-alone executables" is to get all the benefits of the default linking/binding. You could leave a v3 test project as a class library which is imported by (and thus linked/bound in the context of) a different entry point. This entry point could spin up its normal infrastructure and then conditionally run the unit tests if that's what's desired. This is something I have planned for but not yet done myself, so there may be some rough patches figuring out whether I have all the right APIs exposed, etc., but for sure this was a planned-for scenario.

**Comment by joshmouch** [comment]
One problem solved: 

My assumption after running an Aspire host and then looking at the environment variables that are modified was that Service Discovery was using those to pass service mappings around.  That's only partially true, though.

Service Discovery didn't work is because `ConfigurationServiceEndPointResolver.ResolveInternal()` uses `IConfiguration` to find the service mappings, and it assumes that `IConfiguration` will be populated with environment variables.

So, I think this means Aspire services require the orchestrated child app to add `IConfigurationBuilder.AddEnvironmentVariables()` or `IConfigurationManager.AddEnvironmentVariables()` or similar.  This is usually the default, so probably not many others came across this.

This seems like an invalid (exceptional) state to me.  It's not a typed contract, but it's a contract.

I think perhaps that if `ConfigurationServiceEndPointResolver` used environment variables to define discoverable services on the orchestrator side, then `ConfigurationServiceEndPointResolver` should throw an exception if it detects IConfiguration doesn't contain those same variables on the orchestrated side. 

That could apply to any orchestrator/orchestrated contract.  (e.g. RedisServiceEndPointResolver)
